core/src/Kokkos_HPX.hpp

```diff
@@ -49,124 +49,76 @@ static_assert(false,
 
 #include <KokkosExp_MDRangePolicy.hpp>
 
-#include <hpx/local/algorithm.hpp>
 #include <hpx/local/barrier.hpp>
 #include <hpx/local/condition_variable.hpp>
 #include <hpx/local/execution.hpp>
 #include <hpx/local/future.hpp>
-#include <hpx/local/init.hpp>
 #include <hpx/local/mutex.hpp>
-#include <hpx/local/runtime.hpp>
 #include <hpx/local/thread.hpp>
 
 #include <Kokkos_UniqueToken.hpp>
 
+#include <iosfwd>
 #include <functional>
-#include <iostream>
 #include <memory>
-#include <sstream>
 #include <type_traits>
 #include <vector>
 
-// There are currently two different implementations for the parallel dispatch
-// functions:
-//
-// - 0: The HPX way. Unfortunately, this comes with unnecessary
-//      overheads at the moment, so there is
-// - 1: The manual way. This uses for_loop, but only spawns one task per worker
-//      thread. This is significantly faster in most cases.
-//
-// In the long run 0 should be the preferred implementation, but until HPX is
-// improved 1 will be the default.
-#ifndef KOKKOS_HPX_IMPLEMENTATION
-#define KOKKOS_HPX_IMPLEMENTATION 1
-#endif
-
-#if (KOKKOS_HPX_IMPLEMENTATION < 0) || (KOKKOS_HPX_IMPLEMENTATION > 1)
-#error "You have chosen an invalid value for KOKKOS_HPX_IMPLEMENTATION"
-#endif
-
-// [note 1]
-//
-// When using the asynchronous backend and independent instances, we explicitly
-// reset the shared data at the end of a parallel task (execute_task). We do
-// this to avoid circular references with shared pointers that would otherwise
-// never be released.
-//
-// The HPX instance holds shared data for the instance in a shared_ptr. One of
-// the pieces of shared data is the future that we use to sequence parallel
-// dispatches. When a parallel task is launched, a copy of the closure
-// (ParallelFor, ParallelReduce, etc.) is captured in the task. The closure
-// also holds the policy, the policy holds the HPX instance, the instance holds
-// the shared data (for use of buffers in the parallel task). When attaching a
-// continuation to a future, the continuation is stored in the future (shared
-// state). This means that there is a cycle future -> continuation -> closure
-// -> policy -> HPX -> shared data -> future. We break this by releasing the
-// shared data early, as (the pointer to) the shared data will not be used
-// anymore by the closure at the end of execute_task.
-//
-// We also mark the shared instance data as mutable so that we can reset it
-// from the const execute_task member function.
-
 namespace Kokkos {
 namespace Impl {
-class thread_buffer {
+class hpx_thread_buffer {
   static constexpr std::size_t m_cache_line_size = 64;
 
-  std::size_t m_num_threads;
-  std::size_t m_size_per_thread;
-  std::size_t m_size_total;
-  char *m_data;
+  std::size_t m_num_threads     = 0;
+  std::size_t m_size_per_thread = 0;
+  std::size_t m_extra_space     = 0;
+  std::size_t m_size_total      = 0;
+  char *m_data                  = nullptr;
```

Comment:
Did you consider making it a `std::unique_ptr<char[]>` so you don't have to do explicit lifetime management?

Reply:
Yep, I didn't want to mix it in this same PR (it was already using a raw pointer), but it's a small change and I can add it here as well.

Reply:
Changed in https://github.com/kokkos/kokkos/pull/5628/commits/a92a1259aaac31962fdc3492237cee8dd20914dc.